### PR TITLE
The Uni keyboard. Add keyboard

### PR DESCRIPTION
**Pull request by petercpark**
Addition of the Uni and its keymaps to the qmk directory.

## Description
The Uni is an unibody split ortholinear keyboard made specifically for stenography to be used in conjunction with Plover.

Add ./keyboards/the_uni and files within the folder.
## Types of Changes
Add keyboard and keymap.

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/the_uni/config.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x2353
```
Comment (drashna) [inline keyboards/the_uni/config.h:23]: This Vender ID is already in use.  Please change it. Above 0x4000 is pretty safe

https://www.the-sz.com/products/usbid/index.php?v=0x2353&p=&n=

File: keyboards/the_uni/config.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#define PRODUCT_ID      0x0001
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    stenokeyboards
+#define PRODUCT         The Uni
+#define DESCRIPTION     Used in conjunction with plover stenography
```
Comment (drashna) [inline keyboards/the_uni/config.h:28]:
```suggestion
```

File: keyboards/the_uni/config.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* key combination for command */
+#define IS_COMMAND() ( \
+    keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
+)
+
+/* prevent stuck modifiers */
+#define PREVENT_STUCK_MODIFIERS
```
Comment (drashna) [inline keyboards/the_uni/config.h:57]:
```suggestion
```

File: keyboards/the_uni/config.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+/* key matrix pins */
+#define MATRIX_ROW_PINS { F4, B2, B6 }
+#define MATRIX_COL_PINS { F5, F6, F7, B1, B3, B5, B4, E6, D7, C6, D4 }
+#define UNUSED_PINS
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCING_DELAY 5
```
Comment (drashna) [inline keyboards/the_uni/config.h:43]:
```suggestion
#define DEBOUNCE 5
```

File: keyboards/the_uni/keymaps/qwerty/keymap.c [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "the_uni.h"
```
Comment (drashna) [inline keyboards/the_uni/keymaps/qwerty/keymap.c:18]:
```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/the_uni/readme.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+​
+A compact unibody split ortholinear keyboard made specifically for stenography.
+​
+
+-   Keyboard Maintainer: [Peter C. Park](https://github.com/petercpark)
+-   Link to Keyboard Files: [The_Uni](https://github.com/petercpark/The_Uni)
+-   Hardware Supported: The Uni v1 and v2 (pro micro or equivalent)
+-   Hardware Availability: [website](https://www.stenokeyboards.com)
+    ​
+    Make example for this keyboard (after setting up your build environment):
```
Comment (drashna) [inline keyboards/the_uni/readme.md:15]: Could you add instructions on how to enter the bootlaoder for the board?

File: keyboards/the_uni/rules.mk [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
Comment (drashna) [inline keyboards/the_uni/rules.mk:41]:
```suggestion
# Bootloader
BOOTLOADER = atmel-dfu
```

File: keyboards/the_uni/rules.mk [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+COMMAND_ENABLE ?= no    # Commands for debug and configuration
+SLEEP_LED_ENABLE ?= no  # Breathing sleep LED during USB suspend
+NKRO_ENABLE ?= yes		# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE ?= no  # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE ?= no
+
+FORCE_NKRO = yes
+EXTRAKEY_ENABLE = no
+MOUSEKEY_ENABLE = no        # Mouse keys(+4700)
+STENO_ENABLE = yes          # Additional protocols for Stenography(+1700), requires VIRTSER
```
Comment (drashna) [inline keyboards/the_uni/rules.mk:58]:
```suggestion
BOOTMAGIC_ENABLE = lite	# Virtual DIP switch configuration
CONSOLE_ENABLE = no	# Console for debug
COMMAND_ENABLE = no    # Commands for debug and configuration
SLEEP_LED_ENABLE = no  # Breathing sleep LED during USB suspend
NKRO_ENABLE = yes		# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
BACKLIGHT_ENABLE = no  # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no
FORCE_NKRO = yes
EXTRAKEY_ENABLE = no
MOUSEKEY_ENABLE = no        # Mouse keys
STENO_ENABLE = yes      
```

File: keyboards/the_uni/the_uni.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+#define THE_UNI_H
```
Comment (drashna) [inline keyboards/the_uni/the_uni.h:19]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/the_uni/config.h (hunk 1 above)
Comment (petercpark) [inline keyboards/the_uni/config.h:23]:
```suggestion
#define VENDOR_ID       0x9000
```

File: keyboards/the_uni/readme.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    ​
```
Comment (petercpark) [inline keyboards/the_uni/readme.md:16]:
```suggestion
    ​To enter the bootloader, short the rst and gnd pins with a conductive material such as a tweezer or a wire. Uni v2 will have reset pads next to the pro micro so you can short these pads to reset the pro micro.
```

File: keyboards/the_uni/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	 	KC_Q, KC_W, KC_E, KC_R, LT(_CHOOSE, KC_T), KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, 
+	 	KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, 
+	 	            KC_1, KC_C, KC_V, KC_N, KC_M, KC_2),
+
+	 [_CHOOSE] = KEYMAP(
+	 	_______, _______, _______, _______, _______, PLOVER, _______, _______, _______, _______, _______, 
+	 	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, 
+	 	            _______, _______, _______, _______, _______, _______),
+*/
+	[_PLOVER] = KEYMAP(
```
Comment (ridingqwerty) [inline keyboards/the_uni/keymaps/default/keymap.c:48]:
```suggestion
	 [_QWERTY] = LAYOUT(
	 	KC_Q, KC_W, KC_E, KC_R, LT(_CHOOSE, KC_T), KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, 
	 	KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, 
	 	            KC_1, KC_C, KC_V, KC_N, KC_M, KC_2),

	 [_CHOOSE] = LAYOUT(
	 	_______, _______, _______, _______, _______, PLOVER, _______, _______, _______, _______, _______, 
	 	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, 
	 	            _______, _______, _______, _______, _______, _______),
*/
	[_PLOVER] = LAYOUT(
```

Prefer `LAYOUT` over `KEYMAP`.  The json already references `LAYOUT` in this PR so the configurator would likely not work as-is.

File: keyboards/the_uni/keymaps/qwerty/keymap.c [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	 	KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, 
+	 	KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, 
+	 	            KC_1, KC_C, KC_V, KC_N, KC_M, KC_2),
+/*
+
+	 [_CHOOSE] = KEYMAP(
+	 	_______, _______, _______, _______, _______, PLOVER, _______, _______, _______, _______, _______, 
+	 	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, 
+	 	            _______, _______, _______, _______, _______, _______),
+	[_PLOVER] = KEYMAP(
```
Comment (ridingqwerty) [inline keyboards/the_uni/keymaps/qwerty/keymap.c:47]:
```suggestion
	 [_QWERTY] = LAYOUT(
	 	KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, 
	 	KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, 
	 	            KC_1, KC_C, KC_V, KC_N, KC_M, KC_2),
/*

	 [_CHOOSE] = LAYOUT(
	 	_______, _______, _______, _______, _______, PLOVER, _______, _______, _______, _______, _______, 
	 	_______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, 
	 	            _______, _______, _______, _______, _______, _______),
	[_PLOVER] = LAYOUT(
```

Prefer `LAYOUT` over `KEYMAP`.

File: keyboards/the_uni/the_uni.h [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "quantum.h"
+
+#define KEYMAP( \
```
Comment (ridingqwerty) [inline keyboards/the_uni/the_uni.h:22]:
```suggestion
#define LAYOUT( \
```

Prefer `LAYOUT` over `KEYMAP`.

**Review by ridingqwerty — CHANGES_REQUESTED** [review]

File: keyboards/the_uni/config.h [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+
+/*force the nkro if it does not work*/
+#define FORCE_NKRO
```
Comment (drashna) [inline keyboards/the_uni/config.h:52]: In general, I recommend against this, as it can cause the keyboard to become unusable before the OS has booted up.

**Review by drashna — APPROVED** [review]

**Review by ridingqwerty — APPROVED** [review]
Looks good, thanks.

**Pull request merged**